src/java/org/apache/cassandra/index/sai/disk/v3/BlockTerms.java

```diff
@@ -0,0 +1,1887 @@
+import org.apache.cassandra.io.tries.IncrementalDeepTrieWriterPageAware;
+import org.apache.cassandra.io.tries.IncrementalTrieWriter;
+import org.apache.cassandra.io.util.FileHandle;
+import org.apache.cassandra.io.util.FileUtils;
+import org.apache.cassandra.utils.Pair;
+import org.apache.cassandra.utils.Throwables;
+import org.apache.cassandra.utils.bytecomparable.ByteComparable;
+import org.apache.cassandra.utils.bytecomparable.ByteSource;
+import org.apache.cassandra.utils.bytecomparable.ByteSourceInverse;
+import org.apache.lucene.store.IndexInput;
+import org.apache.lucene.store.IndexOutput;
+import org.apache.lucene.util.BytesRef;
+import org.apache.lucene.util.BytesRefArray;
+import org.apache.lucene.util.BytesRefBuilder;
+import org.apache.lucene.util.Counter;
+import org.apache.lucene.util.FutureArrays;
+import org.apache.lucene.util.LongBitSet;
+import org.apache.lucene.util.mutable.MutableValueLong;
+import org.apache.lucene.util.packed.DirectWriter;
+
+import static org.apache.cassandra.index.sai.disk.format.IndexComponent.BLOCK_POSTINGS;
+import static org.apache.cassandra.index.sai.disk.format.IndexComponent.BLOCK_TERMS_DATA;
+import static org.apache.cassandra.index.sai.disk.format.IndexComponent.BLOCK_TERMS_INDEX;
+import static org.apache.cassandra.index.sai.disk.format.IndexComponent.BLOCK_UPPER_POSTINGS;
+import static org.apache.cassandra.index.sai.disk.v1.trie.TrieTermsDictionaryReader.trieSerializer;
+import static org.apache.cassandra.index.sai.disk.v2.sortedterms.SortedTermsWriter.TERMS_DICT_BLOCK_MASK;
+import static org.apache.cassandra.index.sai.disk.v2.sortedterms.SortedTermsWriter.TERMS_DICT_BLOCK_SHIFT;
+import static org.apache.cassandra.index.sai.disk.v2.sortedterms.SortedTermsWriter.copyBytes;
+import static org.apache.cassandra.index.sai.disk.v3.SegmentNumericValuesWriter.BLOCK_SIZE;
+import static org.apache.cassandra.index.sai.disk.v3.SegmentNumericValuesWriter.MONOTONIC_BLOCK_SIZE;
+
+/**
+ * Block terms + postings disk index data structure.
+ *
+ * Each unique posting block min term is indexed into a trie, postings of block size, and optionally an order map when row ids are out of order.
+ *
+ * A string of posting blocks with the same min term are placed into a multi-block spanning posting list.
+ *
+ * Posting block min terms are indexed into a trie.  The payload consists of min and max posting block ids encoded as 2 integers into a long.
+ *
+ * Raw bytes are prefix encoded in blocks of 16 for fast(er) random access than the posting block default size of 1024.
+ *
+ * Block terms reader uses no significant heap space, all major data structures are disk based.
+ *
+ * A point is defined as a term + rowid monotonic id.
+ *
+ * The default postings block size is 1024 or 1k.
+ *
+ * All major data structures are on disk thereby enabling many indexes.
+ */
+public class BlockTerms
+{
+    public static final int DEFAULT_POSTINGS_BLOCK_SIZE = 1024;
```

Comment:
Would it make sense to make this configurable with a system property for testing? I'm thinking of fallout testing with different block sizes.